ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala

```diff
@@ -62,41 +62,53 @@ private[dao] final class TransactionsReader(
       filter: FilterRelation,
       verbose: Boolean,
   ): Source[(Offset, GetTransactionsResponse), NotUsed] = {
-    val events: Source[EventsTable.Entry[Event], NotUsed] =
-      PaginatingAsyncStream.streamFrom((startExclusive, Option.empty[Int]), getOffset[Event]) {
-        case (prevOffset, prevNodeIndex) =>
-          val query =
-            EventsTable
-              .preparePagedGetFlatTransactions(sqlFunctions)(
-                startExclusive = prevOffset,
-                endInclusive = endInclusive,
-                filter = filter,
-                pageSize = pageSize,
-                previousEventNodeIndex = prevNodeIndex,
-              )
-              .withFetchSize(Some(pageSize))
-          val rawEventsFuture =
-            dispatcher.executeSql(dbMetrics.getFlatTransactions) { implicit connection =>
-              query.asVectorOf(EventsTable.rawFlatEventParser)
-            }
-          rawEventsFuture.flatMap(
-            rawEvents =>
-              Timed.future(
-                future = Future.traverse(rawEvents)(deserializeEntry(verbose)),
-                timer = dbMetrics.getFlatTransactions.translationTimer,
+
+    val eventsRangeF: Future[EventsRange[Long]] = dispatcher.executeSql(dbMetrics.getRowIdRange) {
+      connection =>
+        EventsRange.eventsRange(EventsRange(startExclusive, endInclusive))(connection)
+    }
+
+    def getEvents(eventsRange0: EventsRange[Long]): Source[EventsTable.Entry[Event], NotUsed] =
+      PaginatingAsyncStream.streamFrom(
+        eventsRange0,
+        getEventsRange[Event](eventsRange0.endInclusive)) { eventsRange1 =>
+        val query =
+          EventsTable
+            .preparePagedGetFlatTransactions(sqlFunctions)(
+              range = eventsRange1,
+              filter = filter,
+              pageSize = pageSize,
             )
+            .withFetchSize(Some(pageSize))
+        val rawEvents =
+          dispatcher.executeSql(dbMetrics.getFlatTransactions) { implicit connection =>
+            query.asVectorOf(EventsTable.rawFlatEventParser)
+          }
+        rawEvents.flatMap(
+          es =>
+            Timed.future(
+              future = Future.traverse(es)(deserializeEntry(verbose)),
+              timer = dbMetrics.getTransactionTrees.translationTimer,
           )
+        )
       }
 
+    val events: Source[EventsTable.Entry[Event], NotUsed] =
+      Source.future(eventsRangeF).flatMapConcat(r => getEvents(r))
+
     groupContiguous(events)(by = _.transactionId)
-      .flatMapConcat { events =>
+      .mapConcat { events =>
```

Comment:
Does this maintain the existing order? I believe we should return transaction in the right order.

Reply:
it keep the order... it is exactly the same thing. The only difference is we don't need to wrap the result collection into akka `Source`.